include/ada/url_aggregator.h

```diff
@@ -30,28 +30,29 @@ namespace ada {
     bool set_href(const std::string_view input);
     bool set_host(const std::string_view input);
     bool set_hostname(const std::string_view input);
+    inline void set_scheme(std::string_view new_scheme) noexcept;
 
-    [[nodiscard]] ada_really_inline ada::url_components get_components() noexcept;
+    inline void copy_scheme(const url_aggregator& u) noexcept;
 
     [[nodiscard]] std::string get_origin() const noexcept;
-    [[nodiscard]] std::string get_href() const noexcept;
-    [[nodiscard]] std::string get_username() const noexcept;
-    [[nodiscard]] std::string get_password() const noexcept;
-    [[nodiscard]] std::string get_port() const noexcept;
-    [[nodiscard]] std::string get_hash() const noexcept;
-    [[nodiscard]] std::string get_host() const noexcept;
-    [[nodiscard]] std::string get_hostname() const noexcept;
-    [[nodiscard]] std::string get_pathname() const noexcept;
-    [[nodiscard]] std::string get_search() const noexcept;
-    [[nodiscard]] std::string get_protocol() const noexcept;
+    [[nodiscard]] const std::string& get_href() const noexcept;
+    [[nodiscard]] std::string_view get_username() const noexcept;
+    [[nodiscard]] std::string_view get_password() const noexcept;
+    [[nodiscard]] std::string_view get_port() const noexcept;
+    [[nodiscard]] std::string_view get_hash() const noexcept;
+    [[nodiscard]] std::string_view get_host() const noexcept;
+    [[nodiscard]] std::string_view get_hostname() const noexcept;
+    [[nodiscard]] std::string_view get_pathname() const noexcept;
+    [[nodiscard]] std::string_view get_search() const noexcept;
+    [[nodiscard]] std::string_view get_protocol() const noexcept;
```

Comment:
Previously these were all `virtual` functions that we, if forgot to implement, was throwing errors. This gave us a sense of security in case we forgot to implement it. I think we should still make these functions virtual.

Reply:
I think you mean that these should be part of url_base, but I disagree. Let me quote my comment above...

> The problem is that ada::url_aggregator and ada::url have vastly different underlying implementation. Forcing both to conform to the same interface, increases the complexity. It also means that changes on one side (e.g., ada::url_aggregator) impact to the other side (ada::url). So tuning one piece of our code might later require that we update both side. It also might introduce artificial limits. For example, it is natural for ada::url to return an std::string when we ask for the href. However, for ada::url_aggregator, it is maybe more natural to just return a const std::string reference. Similarly, when asking for the url_components, it is natural to ask for a new instance when query ada::url but more natural to get a const reference when querying ada::url_aggregator.


In this instance, we want a different interface for ada::url and ada::url_aggregator. Moving this to ada::url_base force both of them to have the same interface which is likely to force us, for example, to create temporary strings. 

Regarding... "This gave us a sense of security in case we forgot to implement it.", let me refer to my comment above...

> Of course, template programming is difficult. The error messages and insane, and so forth. However, it is an internal concern: users do not need to be concerned about how we implemented the routines. We get nasty programming errors, and failing autocompletion, but the users do not. They either deal with an ada::url_aggregator instance or an ada::url instance. They don't need to know that ada::url_base exists, and they don't need to know that we relied on templates to build our code.